File: BancoSangre/Controllers/DonantesController.cs

using System;
using System.Collections.Generic;
using System.Data.Entity;
using System.Data.Entity.Infrastructure;
using System.Globalization;
using System.Linq;
using System.Web.Mvc;
using BancoSangre.Models;

namespace BancoSangre.Controllers
{
	public class DonantesController : Controller
	{
		private readonly bancosangreEntities _db;

		public DonantesController()
		{
			_db = new bancosangreEntities();
			_db.Configuration.ProxyCreationEnabled = false;
		}

		#region Api

		[HttpGet]
		[Authorize]
		public ActionResult ObtenerDonantes()
		{
			var donantes = _db.Donante.Include(t => t.TipoDocumento).Include(t => t.Localidad).Include(t => t.Provincia).Include(t => t.EstadoDonante)
				.Include(t => t.Donacion)
                .Select(x => new
				{
					x.IdDonante,
					x.TipoDocumento.DescripcionTipoDoc,
					x.NroDoc,
					x.Apellido,
					x.Nombre,
					x.IdGrupoFactor,
					x.GrupoFactor.DescripcionGrupoFactor,
					x.Localidad.NombreLocalidad,
					FechaUltimaDonacion = x.Donacion.Any() ? (DateTime?)x.Donacion.OrderByDescending(w => w.Fecha).FirstOrDefault().Fecha : null,
					x.IdEstadoDonante,
					x.EstadoDonante.DescripcionEstado,
					x.DiferidoHasta
				})
                .OrderByDescending(d => d.FechaUltimaDonacion).ThenBy(x => x.Apellido).ThenBy(x => x.Nombre)
                .ToList();

			if (!donantes.Any())
				return Json(null, JsonRequestBehavior.AllowGet);

			var json = new
			{
				cantidad = donantes.Count,
				data = donantes
			};

			return Json(json, JsonRequestBehavior.AllowGet);
		}

		[HttpPost]
		[Authorize]
		public ActionResult RemoverDonante(long idDonante)
		{
			try
			{
				var actual = _db.Donante
					.FirstOrDefault(x => x.IdDonante == idDonante);
				if (actual == null)
					return Json("El donante que intenta remover no existe en nuestra Base de Datos.", JsonRequestBehavior.AllowGet);

				_db.Donante.Remove(actual);
				_db.SaveChanges();
				return Json(true, JsonRequestBehavior.AllowGet);
			}
			catch (DbUpdateException)
			{
				return Json("El donante no puede ser removido porque ya ha realizado donaciones.", JsonRequestBehavior.AllowGet);
			}
			catch (Exception)
			{
				return Json("Se produjo un error al intentar remover el donante.", JsonRequestBehavior.AllowGet);
			}
		}

		[HttpGet]
		[Authorize]
		public ActionResult ObtenerDonante(long idDonante)
		{

			var donante = _db.Donante.Include(t => t.TipoDocumento).Include(t => t.Localidad).Include(t => t.Provincia).Include(t => t.EstadoDonante)
                .Include(t => t.Donacion).FirstOrDefault(x => x.IdDonante == idDonante);

			var donanteJson = new
			{
				IdDonante = idDonante,
				IdTipoDoc = donante == null ? -1 : donante.IdTipoDoc,
				NroDoc = donante == null ? "" : donante.NroDoc.ToString(),
				Apellido = donante == null ? "" : donante.Apellido,
				Nombre = donante == null ? "" : donante.Nombre,
				IdGrupoFactor = donante == null ? -1 : donante.IdGrupoFactor,
				Domicilio = donante == null ? "" : donante.Domicilio,
				IdProvincia = donante == null ? -1 : donante.IdProvincia,
				IdLocalidad = donante == null ? -2 : donante.IdLocalidad,
				IdEstadoDonante = donante == null ? -1 : donante.IdEstadoDonante,
				DiferidoHasta = donante == null ? null : donante.DiferidoHasta,
				LugarNacimiento = donante == null ? "" : donante.LugarNacimiento,
				OtraLocalidad = "",
				Edad = "",
				Fecha = donante == null ? DateTime.Now : donante.Fecha,
				FechaNacimiento = donante == null ? null : donante.FechaNacimiento,
				Telefono = donante == null ? "" : donante.Telefono,
				Ocupacion = donante == null ? "" : donante.Ocupacion,
                FechaUltimaDonacion = donante != null && donante.Donacion.Any() ? (DateTime?)donante.Donacion.OrderByDescending(w => w.Fecha).FirstOrDefault().Fecha : null,
            };

			if (idDonante != 0 && donante == null)
				return Json(null, JsonRequestBehavior.AllowGet);

			var tiposDocumentos = _db.TipoDocumento.Select(x => new {x.IdTipoDoc, x.DescripcionTipoDoc}).ToList();
			tiposDocumentos.Add(new { IdTipoDoc = -1, DescripcionTipoDoc = "Seleccione Tipo Doc" });
			var gruposFactores = _db.GrupoFactor.Select(x => new { x.IdGrupoFactor, x.DescripcionGrupoFactor }).ToList();
			gruposFactores.Add(new { IdGrupoFactor = -1, DescripcionGrupoFactor = "Seleccione Grupo y Factor" });
			var provincias = _db.Provincia.Select(x => new { x.IdProvincia, x.NombreProvincia }).ToList();
			provincias.Add(new { IdProvincia = -1, NombreProvincia = "Seleccione Provincia" });
			var localidades = ObtenerLocalidades(donanteJson.IdProvincia);
			var estadosDonantes = _db.EstadoDonante.Select(x => new { x.IdEstadoDonante, x.DescripcionEstado }).ToList();
			estadosDonantes.Add(new { IdEstadoDonante = -1, DescripcionEstado = "Seleccione Estado" });

			var json = new
			{
				data = new
				{
					Donante = donanteJson,
					TiposDocumentos = tiposDocumentos.OrderBy(x => x.IdTipoDoc),
					GruposFactores = gruposFactores.OrderBy(x => x.IdGrupoFactor),
					Provincias = provincias.OrderBy(x => x.IdProvincia),
					Localidades = localidades.OrderBy(x => x.IdLocalidad),
					EstadosDonantes = estadosDonantes.OrderBy(x => x.IdEstadoDonante)
				}
			};

			return Json(json, JsonRequestBehavior.AllowGet);
		}

		[Authorize]
		private List<ItemLocalidad> ObtenerLocalidades(int? idProvincia)
		{
            var localidades = _db.Localidad.Where(s => s.IdProvincia == idProvincia || s.IdLocalidad == -1).OrderBy(x => x.NombreLocalidad)
				.Select(x => new ItemLocalidad { IdLocalidad = x.IdLocalidad, NombreLocalidad = x.NombreLocalidad }).ToList();
            localidades.Add(new ItemLocalidad { IdLocalidad = -2, NombreLocalidad = "Seleccione Localidad" });
            return localidades.OrderBy(x => x.IdLocalidad).ToList();
        }
		
		[HttpGet]
		[Authorize]
		public ActionResult TraerLocalidades(int? idProvincia)
		{
			return Json(ObtenerLocalidades(idProvincia), JsonRequestBehavior.AllowGet);
		}

		[HttpPost]
		[Authorize]
		public ActionResult GuardarDonante(Donante donante, string otraLocalidad)
		{
			try
			{
				var idDonante = donante.IdDonante;

				//Otra Localidad
				var idProvincia = donante.IdProvincia;

				if (donante.IdLocalidad == -1)
				{
					var nombreOtraLocalidad = CultureInfo.CurrentCulture.TextInfo.ToTitleCase(otraLocalidad.Trim().ToLower());
					_db.Localidad.Add(new Localidad { NombreLocalidad = nombreOtraLocalidad, IdProvincia = idProvincia });
					_db.SaveChanges();

					donante.IdLocalidad = _db.Localidad.Single(x => x.NombreLocalidad == otraLocalidad && x.IdProvincia == idProvincia).IdLocalidad;
				}

				//Donante
				if (donante.IdDonante == 0)
				{
					//Insertar
					donante.Fecha = DateTime.Now;

					_db.Donante.Add(donante);
					_db.SaveChanges();

					idDonante = _db.Donante.First(x => x.Apellido == donante.Apellido && x.Nombre == donante.Nombre
						&& x.IdTipoDoc == donante.IdTipoDoc && x.NroDoc == donante.NroDoc).IdDonante;
				}
				else
				{
					//Actualizar
					var donanteActual = _db.Donante.Find(donante.IdDonante);
					if (donanteActual == null)
						throw new Exception();

					donanteActual.Nombre = donante.Nombre;
					donanteActual.DiferidoHasta = donante.DiferidoHasta;
					donanteActual.Domicilio = donante.Domicilio;
					donanteActual.Fecha = donante.Fecha;
					donanteActual.Apellido = donante.Apellido;
					donanteActual.FechaNacimiento = donante.FechaNacimiento;
					donanteActual.IdEstadoDonante = donante.IdEstadoDonante;
					donanteActual.IdGrupoFactor = donante.IdGrupoFactor;
					donanteActual.IdLocalidad = donante.IdLocalidad;
					donanteActual.IdProvincia = donante.IdProvincia;
					donanteActual.LugarNacimiento = donante.LugarNacimiento;
					donanteActual.IdTipoDoc = donante.IdTipoDoc;
					donanteActual.NroDoc = donante.NroDoc;
					donanteActual.Telefono = donante.Telefono;
					donanteActual.Ocupacion = donante.Ocupacion;
				}

				_db.SaveChanges();

				var json = new
				{
					data = idDonante,
					resultado = true
				};

				return Json(json, JsonRequestBehavior.AllowGet);
			}
			catch (Exception ex)
			{
				var json = new
				{
					data = ex.Message,
					resultado = false
				};

				return Json(json, JsonRequestBehavior.AllowGet);
			}
		}

		#endregion

		// GET: Donantes
		[Authorize]
		public ActionResult Grilla()
		{
			return View();
		}

		// Recibe idDonante en 0 para diferenciar Ingresar y Editar de Grilla
		[Authorize]
		public ActionResult Ingresar(int? idDonante)
		{
			return View();
		}

		[Authorize]
		public ActionResult Editar(int? idDonante)
		{
			return View();
		}

		protected override void Dispose(bool disposing)
		{
			if (disposing)
			{
				_db.Dispose();
			}
			base.Dispose(disposing);
		}

	}

	internal class ItemLocalidad
	{
		public int IdLocalidad { get; set; }
		public string NombreLocalidad { get; set; }
	}
}

File: BancoSangre/app/controllers/donantesController.js
﻿"use strict";
app.controller("donantesController", function ($scope, donantesRepositorio, modalServicio, $window) {
    init();

    function obtenerParametroPorNombre(nombre) {
        var url = window.location.href;
        nombre = nombre.replace(/[\[\]]/g, "\\$&");
        var regex = new RegExp("[?&]" + nombre + "(=([^&#]*)|&|#|$)"),
            results = regex.exec(url);
        if (!results) return null;
        if (!results[2]) return "";
        return decodeURIComponent(results[2].replace(/\+/g, " "));
    }

    function init() {
        $scope.idDonante = obtenerParametroPorNombre("idDonante");
        if ($scope.idDonante === null) {
            $scope.donacion = obtenerParametroPorNombre("accion") === "donacion";
            $scope.textoBotonEdicion = !$scope.donacion ? 'Editar' : 'Ingresar donación';
            recuperarFiltros();
            configPaginacion();
            obtenerDonantes();
        } else {
            obtenerDonante();
        }
    }

    function configPaginacion() {
        $scope.infoPagina = {
            pagina: 1,
            itemsPorPagina: 100,
            reversa: false,
            totalItems: 0
        };
    }

    function recuperarFiltros() {
        if (localStorage.getItem("busqApellido") != null) $scope.busqApellido = localStorage.getItem("busqApellido");
        if (localStorage.getItem("busqNombre") != null) $scope.busqNombre = localStorage.getItem("busqNombre");
        if (localStorage.getItem("busqGrupo") != null) $scope.busqGrupo = localStorage.getItem("busqGrupo");
        if (localStorage.getItem("busqLocalidad") != null) $scope.busqLocalidad = localStorage.getItem("busqLocalidad");
        if (localStorage.getItem("busqEstado") != null) $scope.busqEstado = parseInt(localStorage.getItem("busqEstado"));
    }

    //Guardar filtros - En _LoginPartial.cshtml removemos las cookies.
    $scope.$watch("busqApellido", function () {
        if ($scope.idDonante === null && $scope.busqApellido !== undefined) localStorage.setItem("busqApellido", $scope.busqApellido); // Sólo para grilla
    });
    $scope.$watch("busqNombre", function () {
        if ($scope.idDonante === null && $scope.busqNombre !== undefined) localStorage.setItem("busqNombre", $scope.busqNombre); // Sólo para grilla
    });
    $scope.$watch("busqGrupo", function () {
        if ($scope.idDonante === null && $scope.busqGrupo !== undefined) localStorage.setItem("busqGrupo", $scope.busqGrupo); // Sólo para grilla
    });
    $scope.$watch("busqLocalidad", function () {
        if ($scope.idDonante === null && $scope.busqLocalidad !== undefined) localStorage.setItem("busqLocalidad", $scope.busqLocalidad); // Sólo para grilla
    });
    $scope.$watch("busqEstado", function () {
        if ($scope.idDonante === null && $scope.busqEstado !== undefined) localStorage.setItem("busqEstado", $scope.busqEstado); // Sólo para grilla
    });

    function obtenerFechaSinFormato(fechaJson) {
        return fechaJson !== undefined && fechaJson !== "" ? new Date(parseInt(fechaJson.replace("/Date(", "").replace(")", ""))) : "";
    }

    function obtenerFechaConFormato(fechaJson) {
        if (fechaJson === undefined || fechaJson === "") return "";

        var fecha = new Date(parseInt(fechaJson.replace("/Date(", "").replace(")", "")));
        var dia = fecha.getDate();
        var mes = fecha.getMonth();
        var anio = fecha.getFullYear();
        return dia + "/" + (mes + 1) + "/" + anio;
    }

    //Grilla

    function obtenerDonantes() {
        if ($scope.busqEstado == null) {
            $scope.busqEstado = -1;
        }
        donantesRepositorio.obtenerDonantes()
            .then(function (result) {
                $scope.donantes = result.data.Donantes;
                $scope.infoPagina.totalItems = result.data.Cantidad;
                $scope.estadosDonantes = result.data.EstadosDonantes;
                for (var i = 0; i < $scope.donantes.length; i++) {
                    $scope.donantes[i].FechaUltimaDonacion = obtenerFechaSinFormato($scope.donantes[i].FechaUltimaDonacion);
                }
            });
    }

    $scope.removerDonante = function (id) {
        donantesRepositorio.remover(id)
            .then(function (result) {
                if (result.data.toString() === "true") {
                    var i;
                    for (i = 0; i < $scope.donantes.length; ++i) {
                        if ($scope.donantes[i].IdDonante === id) {
                            $scope.donantes.splice(i, 1);
                            break;
                        }
                    }
                    modalServicio.open("success", "El Donante ha sido removido correctamente.");
                }
                else {
                    modalServicio.open("danger", result.data);
                }
            });
    };

    $scope.verEstado = function (idEstadoDonante, descripcionEstado, diferidoHasta, idDonante) {
        if (idEstadoDonante === 2 || idEstadoDonante === 3) {
            donantesRepositorio.obtenerCausalesDeRechazo(idDonante).then(function (result) {
                if (result.data.Razones) {
                    modalServicio.open("info", descripcionEstado + ": " + result.data.Razones);
                } else {
                    modalServicio.open("info", descripcionEstado + ". No se han podido recuperar las razones.");
                }
            });
        } else {
            var textoSecundario = "";
            if (idEstadoDonante === 3 && diferidoHasta != null)
                textoSecundario = " hasta " + new Date(diferidoHasta.match(/\d+/)[0] * 1).toLocaleDateString() + ".";
            modalServicio.open("info", descripcionEstado + textoSecundario);
        }
    };

    $scope.obtenerClaseBoton = function () {
        return "btn btn-primary glyphicon glyphicon-" + ($scope.donacion ? "tint" : "pencil");
    };

    $scope.ordenarGrilla = function (columna) {
        $scope.ordenInvertido = $scope.ordenarPorCampo === columna && !$scope.ordenInvertido;
        $scope.ordenarPorCampo = columna;
    };

    $scope.filtroEstado = function (donante) {
        return ($scope.busqApellido == null || donante.Apellido.toLowerCase().includes($scope.busqApellido.toLowerCase()))
            && ($scope.busqNombre == null || donante.Nombre.toLowerCase().includes($scope.busqNombre.toLowerCase()))
            && ($scope.busqGrupo == null || donante.DescripcionGrupoFactor.toLowerCase().includes($scope.busqGrupo.toLowerCase()))
            && ($scope.busqLocalidad == null || donante.NombreLocalidad.toLowerCase().includes($scope.busqLocalidad.toLowerCase()))
            && ($scope.busqEstado === -1 || donante.IdEstadoDonante === $scope.busqEstado);
    };

    $scope.imprimirListado = function () {
        var html = "";
        $("link").each(function () {
            if ($(this).attr("rel").indexOf("stylesheet") !== -1) {
                html += '<link rel="stylesheet" href="' + $(this).attr("href") + '" />';
            }
        });

        var contenidoDiv = $("#listadoDonantes").html();

        contenidoDiv = contenidoDiv.replace(/<th id="thAcciones" style="color: #5a5a5a;"/g, '<th id="thAcciones" style="color: #5a5a5a; visibility:hidden;"')
            .replace(/<td id="tdAcciones" style="white-space: nowrap"/g, '<td id="tdAcciones" style="white-space: nowrap; visibility:hidden;"');

        html += '<body onload="window.focus(); window.print()" style="font-size: 11px !important;"> ' + contenidoDiv + "</body>";
        var w = window.open("", "print");
        if (w) {
            w.document.write(html);
            w.document.close();
        }
    };

    //Creación y edición

    function obtenerDonante() {
        donantesRepositorio.obtenerDonante($scope.idDonante)
            .then(function (result) {
                $scope.donante = result.data.data.Donante;
                if ($scope.donante.FechaNacimiento != null)
                    $scope.calcularEdad(obtenerFechaSinFormato($scope.donante.FechaNacimiento));

                if ($scope.donante.FechaNacimiento != null)
                    $scope.donante.FechaNacimiento = obtenerFechaConFormato($scope.donante.FechaNacimiento);
                if ($scope.donante.DiferidoHasta != null)
                    $scope.donante.DiferidoHasta = obtenerFechaConFormato($scope.donante.DiferidoHasta);
                if ($scope.donante.FechaUltimaDonacion != null)
                    $scope.donante.FechaUltimaDonacion = obtenerFechaConFormato($scope.donante.FechaUltimaDonacion);

                $scope.tiposDocumentos = result.data.data.TiposDocumentos;
                $scope.gruposFactores = result.data.data.GruposFactores;
                $scope.provincias = result.data.data.Provincias;
                $scope.localidades = result.data.data.Localidades;
                $scope.estadosDonantes = result.data.data.EstadosDonantes;
                cargarCalendarios();
            });
    }

    $scope.calcularEdad = function (fecha) {
        var fechaNac = fecha != null ? fecha : $("#calendarioFechaNacimiento").datepicker("getDate");

        if (fechaNac != null) {
            var edad = "";
            var hoy = new Date(Date.now());
            var anioHoy = hoy.getFullYear();
            var mesHoy = hoy.getMonth();
            var diaHoy = hoy.getDate();
            var anioNac = fechaNac.getFullYear();
            var mesNac = fechaNac.getMonth();
            var diaNac = fechaNac.getDate();
            var dif = anioHoy - anioNac;
            if (mesNac > mesHoy) dif--;
            else {
                if (mesNac === mesHoy) {
                    if (diaNac > diaHoy) dif--;
                }
            }
            if (dif !== -1)
                edad = dif;

            $scope.donante.Edad = edad;
        }
    };

    function cargarCalendarios() {
        $.datepicker.setDefaults($.datepicker.regional['es']);

        $("#calendarioFechaNacimiento").datepicker({
            changeMonth: true,
            changeYear: true,
            dateFormat: "dd/mm/yy",
            yearRange: "-90:c"
        }).mask("99/99/9999");

        $("#calendarioDiferidoHasta").datepicker({
            changeMonth: true,
            changeYear: true,
            dateFormat: "dd/mm/yy",
            yearRange: "c:c+10"
        }).mask("99/99/9999");
    }

    $scope.obtenerLocalidades = function () {
        $scope.donante.IdLocalidad = -2;

        donantesRepositorio.obtenerLocalidades($scope.donante.IdProvincia)
            .then(function (result) {
                $scope.localidades = result.data;
            });
    };

    function crearDonante() {
        return {
            IdDonante: $scope.donante.IdDonante,
            IdTipoDoc: $scope.donante.IdTipoDoc,
            NroDoc: $scope.donante.NroDoc,
            Apellido: $scope.donante.Apellido,
            Nombre: $scope.donante.Nombre,
            IdGrupoFactor: $scope.donante.IdGrupoFactor,
            Domicilio: $scope.donante.Domicilio,
            IdProvincia: $scope.donante.IdProvincia,
            IdLocalidad: $scope.donante.IdLocalidad,
            IdEstadoDonante: $scope.donante.IdEstadoDonante,
            DiferidoHasta: $("#calendarioDiferidoHasta").datepicker("getDate"),
            LugarNacimiento: $scope.donante.LugarNacimiento,
            Telefono: $scope.donante.Telefono,
            FechaNacimiento: $("#calendarioFechaNacimiento").datepicker("getDate"),
            Fecha: $scope.donante.Fecha,
            Ocupacion: $scope.donante.Ocupacion,
            FechaUltimaDonacion: $scope.donante.FechaUltimaDonacion,
            CausasIngresadasRechazoDiferido: $scope.donante.CausasIngresadasRechazoDiferido
        };
    }

    $scope.guardar = function (crearCuestionario) {
        if ($scope.donante.IdTipoDoc === -1 || $scope.donante.NroDoc === "" || $scope.donante.Apellido === "" || $scope.donante.Nombre === "" || $scope.donante.IdGrupoFactor === -1 ||
            $scope.donante.IdProvincia === -1 || $scope.donante.IdLocalidad === -2 || $scope.donante.IdLocalidad === -1 && $scope.donante.OtraLocalidad === "" ||
            $scope.donante.IdEstadoDonante === -1 || $scope.localidadExistente)
        {
            $scope.validar = true;
            return;
        }

        var donante = crearDonante();
        donantesRepositorio.guardar(donante, $scope.donante.OtraLocalidad)
            .then(function (result) {
                if (result.data.resultado) {
                    modalServicio.open("success", "El donante se ha guardado con éxito.");
                    if (crearCuestionario)
                        $window.location.href = "/Cuestionarios/Cuestionario?idDonante=" + result.data.data + "&accion=crear";
                    else
                        $window.location.href = "/Donantes/Grilla";
                }
                else {
                    modalServicio.open("danger", result.data.data);
                }
            });
    };

    $scope.$watch('donante.OtraLocalidad', function () {
        if ($scope.localidades !== undefined) {
            for (var i = 0; i < $scope.localidades.length; i++) {
                if ($scope.donante.OtraLocalidad === $scope.localidades[i].NombreLocalidad) {
                    $scope.localidadExistente = true;
                    return;
                }
            }
        }
        $scope.localidadExistente = false;
    });

});